src/parser.js

```diff
@@ -0,0 +1,189 @@
+// [VexFlow](http://vexflow.com) - Copyright (c) Mohit Muthanna 2010.
+//
+// A generic text parsing class for VexFlow.
+
+import { Vex } from './vex';
+
+// To enable logging for this class. Set `Vex.Flow.Parser.DEBUG` to `true`.
+function L(...args) { if (Parser.DEBUG) Vex.L('Vex.Flow.Parser', args); }
+
+export class X extends Error {
+  constructor(message, data) {
+    super(message);
+    this.message = message;
+    this.data = data;
+    this.name = 'Parser';
+  }
+}
+
+// Converts parser results into an easy to reference list that can be
+// used in triggers.
+function flattenMatches(results) {
+  if (results.matchedString !== undefined) return results.matchedString;
+  if (results.results) return flattenMatches(results.results);
+  if (results.length === 1) return flattenMatches(results[0]);
+  if (results.length === 0) return null;
+  return results.map(flattenMatches);
+}
+
+// This is the base parser class. Given an arbitrary context-free grammar, it
+// can parse any line and execute code when specific rules are met (e.g.,
+// when a string is terminated.)
+export class Parser {
+  // For an example of a simple grammar, take a look at tests/parser_tests.js or
+  // the EasyScore grammar in easyscore.js.
+  constructor(grammar) {
+    this.grammar = grammar;
+  }
+
+  // Parse `line` using current grammar. Returns {success: true} if the
+  // line parsed correctly, otherwise returns `{success: false, errorPos: N}`
+  // where `errorPos` is the location of the error in the string.
+  parse(line) {
+    this.line = line;
+    this.pos = 0;
+    this.errorPos = -1;
+    const results = this.expect(this.grammar.begin());
+    results.errorPos = this.errorPos;
+    return results;
+  }
+
+  matchFail(returnPos) {
+    if (this.errorPos === -1) this.errorPos = this.pos;
+    this.pos = returnPos;
+  }
+
+  matchSuccess() {
+    this.errorPos = -1;
+  }
+
+  // Look for `token` in this.line[this.pos], and return success
+  // if one is found. `token` is specified as a regular expression.
+  matchToken(token, noSpace = false) {
+    const re = noSpace ? new RegExp('^((' + token + '))') : new RegExp('^((' + token + ')\\s*)');
```

Comment:
`re` is a pretty standard abbreviation for `regular expression` (across many languages.)